design/design-generic-references.md

```diff
@@ -0,0 +1,211 @@
+      group: ec2.aws.crossplane.io
+      version: v1alpha1
+      resource: vpcs
+      name: main-vpc
+      fieldPath: spec.forProvider.cidrBlock
+    toFieldPath: spec.cidrBlock
+```
+```yaml
+# A generic reference to a namespaced Kubernetes resource.
+spec:
+  externalValues:
+  - fromObject:
+      version: v1
+      resource: configmaps
+      name: common-settings
+      namespace: crossplane-system
+      fieldPath: data.region
+    toFieldPath: spec.forProvider.region
+```
+
+The syntax of `fieldPath` and `toFieldPath` fields will be similar to what we
+use in Composition, in line with [Kubernetes API Conventions](https://github.com/kubernetes/community/blob/744e270/contributors/devel/sig-architecture/api-conventions.md#selecting-fields).
+
+The provider pods will resolve the references via managed reconciler, hence they
+will need at least read-only permissions for any resource that's been referenced
+by an instance. Currently, every provider [has read access](https://github.com/crossplane/crossplane/blob/d8f57a8/internal/controller/rbac/provider/roles/roles.go#L66)
+to all of its own kinds, `Secret`s, `ConfigMap`s, `Lease`s and `Event`s in all
+namespaces. In order to reference resources other than the ones listed, users will
+have to create necessary RBAC resources manually to grant the permissions.
+
+### Implementation
+
+The new field will exist in all managed resources and the cross-resource reference
+fields will stay as is and we'll continue adding them wherever possible. They will
+still be the main way to refer to another resource. The generic references will
+help in cases where that's not possible or not implemented yet.
+
+The API struct and the generic resolver will live in Crossplane Runtime and be
+integrated into managed reconciler. Providers will only have to update their
+runtime dependency and regenerate their CRDs to get the feature in.
+
+One of the caveats with following the new conventions is that controller-runtime
+utilities mostly assume that you have Group-Version-Kind of the resource you're
+operating with. In order to use Group-Version-Resource, we will make use of
+[`EquivalentResourceRegistry`](https://github.com/kubernetes/apimachinery/blob/bf1bfd9/pkg/runtime/mapper.go#L44)
+object as a translator between the two, implemented as an alternative to
+[`MustCreateObject`](https://github.com/crossplane/crossplane-runtime/blob/406fe0b/pkg/resource/resource.go#L145).
+After getting a `runtime.Object`, all controller-runtime utilities should be
+available to use.
+
+The resolver will make use of `fieldpath` library we're using for composition to
+get and set the values unknown types and once it is done, an update request will
+be issued. In line with cross-resource references, if the field on the referenced
+object, pointed by `toFieldPath`, already has a value then the resolver will
+skip it, meaning it will resolve only once and users will need to delete that
+value for resolver to set it again.
```

Comment:
We have a lot of feedback on https://github.com/crossplane/crossplane-runtime/issues/250 requesting that references be resolved on every reconcile, i.e. that the reference take precedence over the underlying field's value.

Reply:
@negz I kind of want either both (CRR and generic refs) or none of the mechanism to resolve at every reconcile so that the resolver behavior is consistent. Do you think both should resolve every time?

Reply:
> I kind of want either both (CRR and generic refs) or none of the mechanism to resolve at every reconcile so that the resolver behavior is consistent.

I agree. Though that said I'm also arguing for there to only be one reference implementation.

> Do you think both should resolve every time?

I think it's a tradeoff. :) I've made my case over in https://github.com/crossplane/crossplane-runtime/issues/250 for the current logic (which I believe to be the common way selectors work upstream) but I can also see that I'm in the minority and that users are finding the current implementation to be unintuitive.